`infrastructure`: Add content for the Processing section (#38)
open-education-hub/methodology: chapters/infrastructure/processing/reading/README.md

# Content Processing

Educational content organized in a Version Control System like `git`, and published to a `GitHub` repository is suitable for collaborative development.
However, it might be a challenging medium for learners to interact with directly.
As a consequence, the content should be presented in a much more friendly and interactive manner.
To that end, the content is processed so that it is deliverable to a publisher (website) in a more approachable way.

The types of content that need to be processed vary with course requirements and require different actions in order to be rendered.
The various types of content are defined and explained in the [Developing and Organizing Content](../../../develop-organize/overview/reading/README.md) chapter.

The rendered output that we recommend is a website generated using a static website generator.

> The usage of [Docusaurus](https://docusaurus.io/) is also recommended, but other static website generators ([Hugo](https://gohugo.io/), [Docsy](https://www.docsy.dev/), [Sphinx](https://www.sphinx-doc.org/en/master/)) can be used.

Reviewer: ```suggestion
The usage of [`Docusaurus`](https://docusaurus.io/) is also recommended, but other static website generators ([`Hugo`](https://gohugo.io/), [`Docsy`](https://www.docsy.dev/), [`Sphinx`](https://www.sphinx-doc.org/en/master/)) can be used.
```

The rendered website can embed materials such as slides and quizzes, and can also refer the end user/learner to the `GitHub` repository from which the content is generated (or to an auxiliary repository) for practical content, such as drills.

Most static website generators have a documented structure that needs to be followed for the website to render.
This can be incompatible with the way content is organized for easier collaboration.

To facilitate content processing with the scope of being published, we have developed the [`Open Education Builder`](https://open-education-hub.github.io/openedu-builder/).

## Open Education Builder

The Open Education Builder is an application that helps organize, transform, and present content.
It is built with a modular architecture in mind, the smallest building block being the `plugin`.

> A `plugin` represents a simple script that transforms the original content in some way.

Reviewer 2: Please be more precise. "in some way" sounds like amateur scientific prose.

A few examples of such plugins are the ones developed for the technologies recommended in our methodology:

- [`RevealMD`](https://open-education-hub.github.io/openedu-builder/plugins/revealmd) - builds multiple slides that are based on `reveal-md`;

> - [`Docusaurus`](https://open-education-hub.github.io/openedu-builder/plugins/docusaurus) - translates the content so that the structures matches the input expectations of Docusaurus.

Reviewer: ```suggestion
- [`Docusaurus`](https://open-education-hub.github.io/openedu-builder/plugins/docusaurus) - translates the content so that the structures match the input expectations of `Docusaurus`.
```

> The plugins are configured through a central configuration file (`config.yaml`), present at the root of your content repository.

Reviewer 2: ```suggestion
The plugins are configured through a central configuration file (`config.yaml`), present at the root of the content repository.
```

The main way to use the Open Education Builder is by customizing the default actions and configuration present in the [`oer-template`](https://github.com/open-education-hub/oer-template) repository.

For configuration options, advanced usage and contributions, please refer to the [documentation](https://open-education-hub.github.io/openedu-builder/) and the [Open Education Builder repository](https://github.com/open-education-hub/openedu-builder)

### Usage examples

The recommended usage of the Open Education Builder is through the provided Docker container.

In this sense, one would extend the provided container by creating a new Dockerfile and using the following line at the beginning:
```Dockerfile
FROM ghcr.io/open-education-hub/openedu-builder:0.5.1
```

> Following, any additional dependencies (e.g. Docusaurus, reveal-md) need to be installed.

Reviewer: ```suggestion
Following, any additional dependencies (e.g. `Docusaurus`, `reveal-md`) need to be installed.
```

In the snippet below, there are 3 steps that will run the `command`, `reveal_embed` and `docusaurus` plugins, in that order, with their respective default input and output directories.

```yaml
stages:
  - make_assets
  - embed_reveal
  - docusaurus

build_dir: /build
output_dir: /output

make_assets:
  plugin: command
# [...]

embed_reveal:
  plugin: reveal_embed
# [...]

docusaurus:
  plugin: docusaurus
# [...]
```

Using content that was built/generated in a previous step can be done by using the full path to that content.
In the example below, we are using content generated in the `make_assets` step as an argument to the `static_assets` option of the `docusaurus` plugin.

```yaml
make_assets:
  plugin: command
  options:
    command: make
    locations:
      - content/chapters/data/lecture
      - content/chapters/io/lecture
      # [...]
    args:
      - all

# [...]

docusaurus:
  plugin: docusaurus
  options:
# [...]
    static_assets:
      - slides/Data: /build/make_assets/content/chapters/data/lecture/_site
      - slides/IO: /build/make_assets/content/chapters/io/lecture/_site
# [...]

```

> #### Docusaurus

Reviewer: ```suggestion
## Docusaurus
```
Shouldn't this be on the same level with Open Edu Builder, or at least not nested under "Usage examples"?

Author: It's a plugin of the Open Edu Builder and the section contains usage examples. I think the level is fine.

Docusaurus requires a directory structure similar to this:
```
docs

> └── Labs

Reviewer: We avoided concrete examples so far and used either mocks or the methodology repo itself. Not sure if this should also be the case here.

Reviewer 3: I think this is fine. We do aim to use real examples when we have them.

        ├── overview.md
        ├── processes.md
        └── threads.md
```
This is different from the structure proposed in the [Developing and Organizing Content](../../../develop-organize/overview/reading/README.md) chapter.
We have developed the `Docusaurus` plugin for the `Open Education Builder` in order to facilitate this reorganization.

The most important option of this plugin is the `structure` option, which defines the order of the chapters and their sections and subsections, together with the source of the content.
A simplified example, taken from the [Operating Systems](https://github.com/open-education-hub/operating-systems) repository, that maps to the directory structure above can be seen below.

```yaml
# [...]
    structure:
      - Lab:
        - Compute:
            path: content/chapters/compute/lab/content
            extra:
              - ../media
              - ../quiz
            subsections:
              - Overview: overview.md
              - Processes: processes.md
              - Threads: threads.md
# [...]
```
A section (such as `Lab` or `Compute` above) can either be followed by a list of subsections (as is the case for `Lab`) or by a few options.

> In the case of the `Compute` section, it has 3 options: `path`, `extra` and `subsections`.
> 
> The `subsections` option specifies the subsections of the current section. These can be pages that will be rendered (as in this example) or other sections.
> 

Reviewer: ```suggestion
In the case of the `Compute` section, it has 3 options:

- The `subsections` option specifies the subsections of the current section.
  These can be pages that will be rendered (as in this example) or other sections.
```
Use a list for these options rather than a paragraph as they are closely related.

---
title: Content Processing
---

### Content Processing

> Content organized in a Version Control System like `git`, and published to a `GitHub` repository is suitable for easier contributions.

Reviewer 2: ```suggestion
Educational content organized in a Version Control System like `git`, and published to a `GitHub` repository is suitable for collaborative development.
However, it is not a good medium for learners to interact with.
```

> In order to deliver it to the end user/learner, this content has to be processed and rendered in an easier to navigate form.

Reviewer 2: ```suggestion
As a consequence, the content should be presented in a much more friendly and interactive manner.
To that end, the content is processed so that it is deliverable to a publisher (website) in a much more approachable way.
```

The types of content that need to be processed vary with course requirements and require different actions in order to be rendered.
These types include, but are not limited to:

- slides
- reading material
- quizes
- various types of media content.

The rendered output that we recommend is a website generated using a static website generator.
The usage of [Docusaurus](https://docusaurus.io/) is also recommended, but other static website generators (Hugo, Docsy, Sphinx) can be used.

> The rendered website can embed materials such as slides and quizes, and can also refer the end user/learner to the `GitHub` repository from which the content is generated (or to an auxiliary repository) for practical content, such as drills.

Reviewer 2: ```suggestion
The rendered website can embed materials such as slides and quizzes, and can also refer the end user/learner to the `GitHub` repository from which the content is generated (or to an auxiliary repository) for practical content, such as drills.
```

Most static website generators have a documented structure that needs to be followed for the website to render.

> This can be incompatible with the way content is organized for easier collaboration.

Reviewer 3: This is a bit unclear. You're saying that the content should be structured in such a way as to fit the publisher?

To facilitate content processing with the scope of being published, we have developed the [`Open Education Builder`](https://open-education-hub.github.io/openedu-builder/).

#### Open Education Builder

The Open Education Builder is an application that helps with the organization, transformation and output of content.
It is built with a modular architecture in mind, the smallest building block being the `plugin`.
A `plugin` represents a simple script that transforms the original content in some way.
A few examples of such plugins are the ones developed for the technologies recommended in our methodology:

- `revealmd` - builds multiple slides that are based on `reveal-md`;

> - `docusaurus` - helps with organization of content in a manner that is compatible with Docusaurus.

Reviewer 2: `helps` is an inadequate term. Maybe "translates the content so that the structures matches the input expectations of Docusaurus".

The plugins are configured through a central configuration file (`config.yaml`), present at the root of your content repository.

The main way to use the Open Education Builder is by customizing the default actions and configuration present in the [`oer-template`](https://github.com/open-education-hub/oer-template) repository.

> For configuration options, advanced usage and contributions, please refer to the [documentation](https://open-education-hub.github.io/openedu-builder/) and the [Open Education Builder repository](https://github.com/open-education-hub/openedu-builder)

Reviewer 3: You should give some examples here. Otherwise, the config file is going to be difficult to use. Ideally, you would have:
- examples of how to use the builder (both manually and as an action)
- examples of how to use `config.yml` and its interaction with the builder.
